storage/badger/payloads.go

```diff
@@ -43,26 +43,25 @@ func (p *Payloads) storeTx(blockID flow.Identifier, payload *flow.Payload) func(
 	// which is not included in current payload but was incorporated in one of previous blocks.
 	// TODO: refactor receipt/results storages to support new type of storing/retrieving where execution receipt
 	// and execution result is decoupled.
-	resultsByID := payload.Results.Lookup()
-	fullReceipts := make([]*flow.ExecutionReceipt, 0, len(payload.Receipts))
-	var err error
-	for _, meta := range payload.Receipts {
-		result, ok := resultsByID[meta.ResultID]
-		if !ok {
-			result, err = p.results.ByID(meta.ResultID)
-			if err != nil {
-				if errors.Is(err, storage.ErrNotFound) {
-					err = fmt.Errorf("invalid payload referencing unknown execution result %v, err: %w", meta.ResultID, err)
-				}
-				return func(*transaction.Tx) error {
+
+	return func(tx *transaction.Tx) error {
+
+		resultsByID := payload.Results.Lookup()
+		fullReceipts := make([]*flow.ExecutionReceipt, 0, len(payload.Receipts))
+		var err error
+		for _, meta := range payload.Receipts {
+			result, ok := resultsByID[meta.ResultID]
+			if !ok {
+				result, err = p.results.ByIDTx(meta.ResultID)(tx)
```

Comment:
Why this has to be in a transaction?

If a receipt's result is not found in the payload, it means the result must have been seen, which means the storage should have it. Since a result once is stored, will never be removed, it must be found.

Reply:
When we bootstrap the state, we insert all blocks/receipts of the sealing segment within the Badger transaction scope. When we read the DB here, if we don't read in the context of the same Badger transaction, it will not see blocks/receipts inserted earlier in the transaction.